lale/grammar.py

```diff
@@ -0,0 +1,115 @@
+from lale.operators import MetaModelOperator, PlannedOperator, Operator, BasePipeline, OperatorChoice, IndividualOp
+from lale.operators import make_choice, make_pipeline
+from lale.lib.lale import NoOp
+from typing import Optional
+
+class Primitive(Operator):
+    """ Abstract operator for non-terminal grammar rules.
+    """
+    def __init__(self, name, value=None):
+        self._name = name
+        self.value = value
+        
+    def _lale_clone(self):
+        pass
+    
+    def has_same_impl(self):
+        pass
+    
+    def is_supervised(self):
+        return False
+
+    def name(self):
+        return self._name
+    
+    def set_name(self, name):
+        self._name = name
+        
+        
+class Grammar(MetaModelOperator):
+    """ Base class for Lale grammars. The for every grammar `g`, rule `g.start` must be defined.
```

Comment:
- Typo: "The for" -> "For"
- Also, would it make sense to add some error checking in the code for the constraint that `g.start` must be defined?

Reply:
We do an `assert hasattr(g, 'start'), "Rule start must be defined"` in `explore` which is the only place where we rely on this assumption.